src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -280,6 +280,9 @@ def _get_installed_candidate() -> Optional[Candidate]:
             # the current dependency graph.
             if not specifier.contains(installed_dist.version, prereleases=True):
                 return None
+            if installed_dist.direct_url or installed_dist.editable:
+                # Defer to _get_ireq.
+                return None
```

Comment:
This is not correct... perhaps we should defer to _get_ireq only when --upgrade is specified.